ldap/servers/plugins/replication/cl5_api.c

```diff
@@ -4294,63 +4172,66 @@ _cl5CheckMissingCSN(const CSN *csn, const RUV *supplierRuv, cldb_Handle *cldb)
 {
-    int rc;
-    void *iterator = NULL;
+    DBLCI_CTX *dblcictx = ctx;
     slapi_operation_parameters op = {0};
-    char *buff;
-    PRInt32 len, wlen;
+    cldb_Handle *cldb = dblcictx->cldb;
     CL5Entry entry = {0};
-
-    PR_ASSERT(prFile && cldb);
-
-    if (slapi_is_loglevel_set(SLAPI_LOG_REPL)) {
-        ruv_dump(cldb->purgeRUV, "clpurgeruv", prFile);
-        ruv_dump(cldb->maxRUV, "clmaxruv", prFile);
+    PRInt32 len, wlen;
+    char *buff;
+    int rc = _cl5CICbInit(key, data, dblcictx);
+    if (rc != DBI_RC_SUCCESS) {
+        return rc;
     }
-    slapi_write_buffer(prFile, "\n", strlen("\n"));
-
     entry.op = &op;
-    rc = _cl5GetFirstEntry(cldb, &entry, &iterator, NULL);
-    while (rc == CL5_SUCCESS) {
-        rc = _cl5Operation2LDIF(&op, cldb->ident, &buff, &len);
-        if (rc != CL5_SUCCESS) {
-            slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
-                          "_cl5ExportFile - Failed to convert operation to ldif\n");
-            operation_parameters_done(&op);
-            break;
-        }
-
-        wlen = slapi_write_buffer(prFile, buff, len);
-        slapi_ch_free((void **)&buff);
+    rc = cl5DBData2Entry(data->data, data->size, &entry, cldb->clcrypt_handle);
+    if (rc != DBI_RC_SUCCESS) {
+        return rc;
+    }
+    rc = _cl5Operation2LDIF(&op, cldb->ident, &buff, &len);
+    if (rc != CL5_SUCCESS) {
+        slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
+                      "_cl5ExportEntry2File - Failed to convert operation to ldif\n");
+    } else {
+        wlen = slapi_write_buffer(dblcictx->exportFile, buff, len);
+        slapi_ch_free_string(&buff);
         if (wlen < len) {
             slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
-                          "_cl5ExportFile - Failed to write to ldif file\n");
+                          "_cl5ExportEntry2File - Failed to write to ldif file\n");
             rc = CL5_SYSTEM_ERROR;
-            operation_parameters_done(&op);
-            break;
         }
-
-        cl5_operation_parameters_done(&op);
-
-        rc = _cl5GetNextEntry(&entry, iterator);
     }
+    operation_parameters_done(&op);
+    return rc;
+}
 
-    cl5_operation_parameters_done(&op);
-
-    if (iterator)
-        cl5DestroyIterator(iterator);
+static int
+_cl5ExportFile(PRFileDesc *prFile, cldb_Handle *cldb)
+{
+    DBLCI_CTX dblcictx = {0};
+    int rc = CL5_SUCCESS;
+    PR_ASSERT(prFile && cldb);
 
-    if (rc != CL5_NOTFOUND) {
-        slapi_log_err(SLAPI_LOG_ERR, repl_plugin_name_cl,
-                      "_cl5ExportFile - Failed to retrieve changelog entry\n");
-    } else {
-        rc = CL5_SUCCESS;
+    if (slapi_is_loglevel_set(SLAPI_LOG_REPL)) {
+        ruv_dump(cldb->purgeRUV, "clpurgeruv", prFile);
+        ruv_dump(cldb->maxRUV, "clmaxruv", prFile);
     }
+    slapi_write_buffer(prFile, "\n", strlen("\n"));
 
-    return rc;
+    dblcictx.seen.nbmax = CL5_PURGE_MAX_LOOKUP_PER_TRANSACTION;
```

Comment:
This is an export, is it done under a TXN ?

Reply:
Interesting point:
   Apparently the old code used read-only txn on lmdb (lmdb API requires a txn to create a cursor) 
   and none with bdb, but the new code is  always using a read-only txn (cf _cl5Iterate)
I have mixed feeling about that:
  I could always change _cl5Iterate to avoid creating a read only txn in bdb case but I wonder if the old code
  could not lead to weird inconsistency because of the lack of txn ...